src/OrchardCore.Modules/OrchardCore.Contents/Workflows/Startup.cs

```diff
@@ -0,0 +1,29 @@
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.ContentManagement.Handlers;
+using OrchardCore.Contents.Workflows.Activities;
+using OrchardCore.Contents.Workflows.Drivers;
+using OrchardCore.Contents.Workflows.Handlers;
+using OrchardCore.Modules;
+using OrchardCore.Workflows.Helpers;
+using OrchardCore.Workflows.Services;
+
+namespace OrchardCore.Contents.Workflows
+{
+    [RequireFeatures("OrchardCore.Workflows")]
+    public class Startup : StartupBase
+    {
+        public override void ConfigureServices(IServiceCollection services)
+        {
+            services.AddActivity<ContentCreatedEvent, ContentCreatedEventDisplay>();
```

Comment:
We'll use you as a guinea pig. I want to replace some uses of DI by options. The idea here is that you create a new `WorkflowOptions` class, with an `Activities` list, and we then configure workflow by adding instances of activities. It also allows any module to remove activities, or reorder them (though we don't care in this case). This way we it's faster as we don't have to "resolve" them, and it's lighter as the DI is not overflowed.

Reply:
I like it.

Reply:
One problem though; the activity classes currently support constructor-injection. If a module has to add an instance, it first would have to resolve its dependencies so the module can then manually instantiate a given activity class.

I have 3 options to deal with this:

1. Don't add an activity instance, but its type. Then the IActivityLibrary (which is a factory) instantiates an activity using `Activator` (since it can't be resolved from the DI container), but use the DI container to resolve each dependency of the activity (using [this helper class](https://github.com/OrchardCMS/OrchardCore/blob/workflows/src/OrchardCore/OrchardCore.Workflows.Abstractions/Helpers/ServiceProviderExtensions.cs)).
1. Allow activities to resolve services they need via the `WorkflowExecutionContext`. Not as nice because it's now less obvious what services a given activity needs.
1. You tell me. ;)

Reply:
Another thing: activities have display drivers. And I think that if we don't register the activities with the container, the same should go for their drivers. If I go with option 1 mentioned above, then I can have  simple dictionary by activity type with a tuple-like class that holds the activity type and driver type. I can then easily instantiate both. However, I would have to either duplicate (or modify) the DisplayManager class, or instantiate it manually, since its constructor is injected with an enumerable of IDisplayDriver instances, that I now need to resolve myself (from the WorkflowOptions).

Reply:
[Here's a commit](https://github.com/OrchardCMS/OrchardCore/pull/1378/commits/2d8d57d148d9f036b458dc7bf4c0cec35713a7c0) that shows what would need to be done.

It's all quite doable, but with this approach we still depend on the service container to resolve dependencies of activities and activity display drivers. Not sure it's worth it.

Reply:
I think we'll need the `WorkflowContext` to expose the `IServiceProvider`, but as you mentioned we still have the issue for display drivers. So I might just do it first for Parts and DisplayDrivers, then later we can apply the same pattern to workflow and other components. So forget about it for now.

Reply:
OK. In that case, I think it's best to for now register everything with the container as before, so that it works the same as with content parts and drivers. Then once you have the right mechanism in place, we apply it to Workflows.